### PR TITLE
Fix bug with creating records on tables without a primary key

### LINKED ISSUE #3246: Insert broken when there's not primary key
## Description

When a table does not have a primary key, attempting to use the UI to create a new record results in the backend throwing an unhandled 500 error.

## Expected behavior

When a `POST` request representing a new empty record is sent for a table with no primary key, the server should respond with a 405 error, since we can't allow creating (or editing) rows on tables with no primary key. The error should be associated with some informational `APIException` stating that you can't insert into tables without a primary key.

The implementer should add an API test to validate their solution.

## To Reproduce

1. Using psql or some other client, create a table without a primary key.
2. In the UI, navigate to that table.
3. Click the `+` to add a record. 
image
4. Observe the 500 response.
5. See the error (via mouse-over) in the UI: 
image
6. See the `tuple index out of range` error in the devtools.

### PR DESCRIPTION
**Pull request by Shraeyas**
Fixes #3246

**Technical details**

- Added condition to check if a primary key exists on table and updated error message to "You cannot insert into tables without a primary key" and status code to 405 if a user tries to insert a new record.
- Added MethodNotAllowedAPIException error class.
- Updated test coverage for validating record creation on tables without primary key .

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by Shraeyas** [comment]
@mathemancer This will fix #3246

I've updated the tests as well. Can you review this?

**Comment by Shraeyas** [comment]
Not sure why the particular test is failing. It is working fine on my local setup.

**Comment by mathemancer** [comment]
@Shraeyas Don't worry about that test failure. We have a problem with the CI pipeline, addressed by #3254

File: mathesar/api/db/viewsets/records.py [hunk 1]
```diff
@@ -145,17 +145,26 @@ def retrieve(self, request, pk=None, table_pk=None):
 
     def create(self, request, table_pk=None):
         table = get_table_or_404(table_pk)
+        primary_key_column_name = None
+        try:
+            primary_key_column_name = table.primary_key_column_name
+        except Exception:
```
Comment (mathemancer) [inline mathesar/api/db/viewsets/records.py:151]: Please check for a more specific `Exception` here.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Overall, pretty good!

I do have one small request; see the specific comment.

File: mathesar/api/db/viewsets/records.py (hunk 1 above)
Comment (Shraeyas) [reply mathesar/api/db/viewsets/records.py]: I've changed it to 'AssertionError' @mathemancer

**Pull request merged**